### ISSUE TITLE
Add filepath relative to runpath

### ISSUE THREAD

**Issue by perolavsvendsen** [equinor/fmu-dataio#644]
## Context
The concept of relative paths are key in FMU. The relative file path can be seen as a identifier of a specific file inside, and across, realizations. The construction of the file path is not necessarily 1:1 with, or even covered by, the other metadata due to ability to modify the relative filepath (`subfolder`, `forcefolder` args). Also, freetext attributes such as `tagname` are included in the filename.

The `file` block in the metadata contains absolute and relative filepaths:

```yaml
file:
  absolute_path: /scratch/user/case/realization-0/iter-0/share/results/file/myfile.xyz
  relative_path: realization-0/iter-0/share/results/file/myfile.xyz # relative to case root
```

## Change proposal
Expand the `file` block to also include the file path relative to `runpath`:

```yaml
file:
  absolute_path: /scratch/user/case/realization-0/iter-0/share/results/file/myfile.xyz
  relative_path: realization-0/iter-0/share/results/file/myfile.xyz # relative to case root
  internal_relative_path: share/results/file/myfile.xyz # relative to runpath # NEW!
```

> [!NOTE]
> The term `internal_relative_path` may not be the best. As we know, naming things is hard. Partly because we don't have the terminology/concepts for this.

## Use case
One usage of the relative file name is as an identifier of all representations of a specific file across e.g. an ensemble of realizations.

Example: A model is set up to export a specific depth surface, `mysurface`. A representation of this surface is then produced for each realization. The result is an array of representations of this surface (or a _distribution_). Other representations can be added, e.g. statistical representations. When creating statistical aggregations, it is vital to consistently be able to identify the representations (realizations) of _this specific surface_. Currently, the only guarantee for this is the combination of `fmu.case.uuid` + `file.relative_path`.

> [!NOTE]
> The relative file path is also used to create a unique identifier for each individual file when stored in Sumo, by hashing `fmu.case.uuid`+`file.relative_path`.

When doing (statistical) aggregations, it is vitally important to ensure that the right selection of data is used as input. On-premise solutions (Python scripts) that do statistical aggregations are based solely on the file path. Cloud solutions (Sumo-powered services) are currently doing the same in combination with `fmu.iteration.uuid`.

The current format of the relative filepath is relative to the `case`, meaning that it includes the `realization` and `iteration` folder levels. It still needs to do this, because of the usage for unique identifiers per file - not all files belong to iterations or realizations. However, when using the relative file path for identifying a specific file across realizations, the consumer needs to crop the realization and iteration information. This creates overhead and makes it difficult to use the underlying search engine in Sumo properly. Therefore, expanding the `file` block with this attribute may be a step in the right direction.

## Alternatives
The root problem is to be able to confidently identify all representations of a specific file, no more, no less. If other metadata attributes can obtain this, that could lead to the same result. This is not the case today.

A key requirement currently imposed is to have 1:1 behavior between /scratch and Sumo when it comes to overwriting. I.e. if a file is overwritten on /scratch, it is _also overwritten on Sumo_. To obtain this with confidence, we currently utilize the file path since we know that two files cannot have the same file path.

> [!WARNING]
> This is technically only true when all files are produced on the same disk system, at the same time. This is, however, not the case since various post-processing services also produce data to FMU cases. However, given unique UUID per file generated in Sumo based on relative file path, the behavior will be the same.

So given that the file path will remain quite central, we can envision that the relative file path is built using known attributes from metadata. To some degree, this is done already, e.g. 

```
{fmu.realization.name}/{fmu.iteration.name}/share/results/{class}/{[subfolder]}/{name}--{tagname}--{[t0]}-{[t1]}.postfix
```

`fmu-dataio` supports the `subfolder` argument, which allows the user to expand the file path with one level. This is required as the current data model is very limited, and does not scale to existing needs. However, the `subfolder` attribute is not included in the metadata. Is is questionable if it _should_, as it will not necessarily make sense without quite deep knowledge of how FMU works internally. This is not to be expected amongst random data consumers.

One could also imagine that we required that the relative filepath is constructed from existing, and known, metadata attributes and that we don't allow customization of the file path. This is however probably not a good idea, as it does not scale well. (We would have to constantly expand the path whenever we get into uniqueness-issues). Also, the relative filepaths are key to current consumption of FMU results so doing this will potentially be very disruptive.

There is a bigger discussion in the background here related to refining the whole file path concept, which will not be fully covered here.

**Comment by perolavsvendsen** [comment]
Perhaps "identifier" is a better term?

```yaml
file:
  absolute_path: /scratch/user/case/realization-0/iter-0/share/results/file/myfile.xyz
  relative_path: realization-0/iter-0/share/results/file/myfile.xyz # relative to case root
  identifier: share/results/file/myfile.xyz # relative to runpath # NEW!
```

🤔

**Comment by perolavsvendsen** [comment]
Side-discussion: Perhaps it would be beneficial to have a "templated" path, with placeholders for the ensemble (iteration) and realization, i.e.:

```yaml
file:
  absolute_path: /scratch/user/case/realization-0/iter-0/share/results/file/myfile.xyz
  relative_path: realization-0/iter-0/share/results/file/myfile.xyz # relative to case root
  identifier: share/results/file/myfile.xyz # relative to runpath # NEW!
  templated_identifier: {realization}/{iteration}/share/results/file/myfile.xyz # NEW!
```

**Comment by perolavsvendsen** [comment]
Discussions with Dynageo:
- They have the same need.
- Would like to have an identifier that defines a specific "thing", regardless of realization, aggregation, etc.

Example:
An FMU workflow is configured to produce a depth surface "My Horizon". This is produced in all realizations. Afterwards, it should be possible to identify this specific depth surface as a set of different representations (real-0, real-1, ..., real-n).

I.e. looks at a surface from FMU as an array/a distribution of individual representations. Need a way to identify this array/collection/distribution/whatever.

Basically, same need as described above.

Today it is possible to identify which objects that belong to the e.g. same realization, but it is not possible (with confidence) to identify which objects that are realizations of the same "thing".

(This is similar to the current `fmu.aggregation.id` in terms of usage.)

**Comment by tnatt** [comment]
**Question** 
Should the `file.identifier` only be added if we are in a **realization** context? If not, what should be the expected behavior of this field for the different run contexts we have?

It would be nice if the `identifier` could be used to link objects representing the same thing, and that we keep it the same across the various context. It could make it easy to get the aggregated ensemble object for a particular file. 

In the **case** context we can't make any guarantees though since these data are uploaded by users. They may have a file in their case folder with the same name as files stored in realizations, and they might not represent the same thing..
Should we perhaps use `None` as identifier value for **case** contexts?

Will having a file on **case/ensemble** level with the same identifier as a **realization** level file create confusion for users when looking for data in `Sumo`?... This could probably be handled gracefully through the `Explorer` by requiring the desired stage?

I guess you are the one that has put most thought into this @perolavsvendsen ? or do any of you in the sudo team have opinions on this @rwiker @equinor-ruaj ? I don't think I see the full picture 🙂 . 

**context="realization"**
```
file:
  absolute_path: /scratch/user/case/realization-0/iter-0/share/results/file/myfile.xyz
  relative_path: realization-0/iter-0/share/results/file/myfile.xyz # relative to case root
  identifier: share/results/file/myfile.xyz 
```
**context="ensemble"**
```
file:
  absolute_path: /scratch/user/case/iter-0/share/results/file/myfile.xyz
  relative_path: iter-0/share/results/file/myfile.xyz # relative to case root
  identifier: share/results/file/myfile.xyz 
```
**or**
```
  identifier: iter-0/share/results/file/myfile.xyz 
```
**context="case"**
```
file:
  absolute_path: /scratch/user/case/share/results/file/myfile.xyz
  relative_path: share/results/file/myfile.xyz # relative to case root
  identifier: share/results/file/myfile.xyz 
```

**Comment by perolavsvendsen** [comment]
This may also be required for (more effective) license partner sharing. We want to give the option to not share all data in an ensemble, but we also don't want single realizations to be left out. (For many reasons, but hot topic right now is how we cache transformed data.)

But we don't really have a mechanism to identify individual files while also ignoring individual realizations. I.e. we don't have a way to identify "all representations of this 'thing'".

(The lack of terminology is killing me here...)

I think, if we had this identifier, and could start relating to a (distribution of) a specific 'thing', we could also change how we present data in the data viewer in Sumo. I.e. we would possibly just show one entry for each distribution, etc.

Feel free to call for a discussion on this if you feel like it!

**Comment by perolavsvendsen** [comment]
Also, I don't know if the `file` block is the best/correct placement for this. Could be that it is better placed in the `fmu` block (we just make it based on some information in the file block right now, since that is, right now, the only reliable thing we have for uniqueness - but that might change).

This is in my head, conceptually, very similar to `fmu.aggregation.id` - which is a `uuid` that identifies which specific aggregation operation this came from, which in turn is essentially an identifier for a set of objects that are different representations of _the same thing_.

Perhaps something like `fmu.object.uuid`?? I don't know...

(But could be that a `uuid` of some sort would be smart, so that we in the future can change how this is actually generated without causing disruption 🤷‍♂)

(Other opinions are probably needed here...)

**Comment by perolavsvendsen** [comment]
I'm probably overcomplicating things 😬 But lets talk 👍

**Comment by perolavsvendsen** [comment]
Offline discussions:
We want to confidently identify the representations of a specific exported data object. The representations are individual realizations of "the same thing" + aggregations/transformations of this.

- We introduce the truncated relative filepath as described above, e.g. `file.relative_to_runpath` (find good name). The intention shall then _not_ be that any data consumer uses this, but:
- We also create a dedicated identifier under the `fmu` block which shall be the _actual_ identifier to be used. This identifier shall confidently identify members of the same distribution/array of representations. I.e. a consumer can ask for objects with a particular value, and in return get all members of a distribution of "the same thing".
- This ID should/could probably/possibly be logically bound to the ensemble (?). I.e. we assume that the consumer knows which ensemble he is looking for. This needs some further discussion...

- In realization-context objects, we set the file path and the uuid based on this.
- In aggregations (ensemble-context), we set the uuid directly (based on the input).
- In case-context, this is meaningless. Either, we don't set it at all - or we just set a random uuid.

To be further described :)

**Comment by perolavsvendsen** [comment]
Contributing towards https://github.com/equinor/fmu-results-activities/issues/33
